docs/proposals/push-metrics.md

```diff
@@ -0,0 +1,221 @@
+`/api/v1/push/prometheus`:
+
+```
+# This is a pod-level metric (it might be used for autoscaling)
+# TYPE http_requests_per_second guage
+http_requests_per_minute{namespace="webapp",pod="frontend-server-a-1"} 20
+http_requests_per_minute{namespace="webapp",pod="frontend-server-a-2"} 5
+http_requests_per_minute{namespace="webapp",pod="frontend-server-b-1"} 25
+
+# This is a service-level metric, which will be stored as frontend_hits_total
+# and restapi_hits_total (these might be used for auto-idling)
+# TYPE hits_total counter
+hits_total{namespace="webapp",service="frontend"} 5000
+hits_total{namespace="webapp",service="restapi"} 6000
+```
+
+This would result in the metrics being available at:
+
+```
+/api/v1/model/namespaces/webapp/pods/frontend-server-a-1/metrics/custom/http_gatherer/http_requests_per_minute
+/api/v1/model/namespaces/webapp/pods/frontend-server-a-2/metrics/custom/http_gatherer/http_requests_per_minute
+/api/v1/model/namespaces/webapp/pods/frontend-server-b-1/metrics/custom/http_gatherer/http_requests_per_minute
+/api/v1/model/namespaces/webapp/metrics/custom/http_gatherer/frontend/hits_total
+/api/v1/model/namespaces/webapp/metrics/custom/http_gatherer/restapi/hits_total
+```
+
+## Discussed Alternatives
+
+A number of alternatives came up during the discussion of this proposal.  They
+are discussed briefly below.  Note that most of these alternatives do not deal
+particularly well with a case where metrics need to come from a source that is
+not running as a pod on the cluster.  While it is expected that many of the
+producers will be running as components on the cluster (e.g. as DaemonSets or
+PetSets), it could still be adventageous to support metrics coming from
+components that are not in the form of pods.
+
+### Writing directly into sinks
+
+This alternative would have producers write directly into sinks in the Heapster
+storage schema, and then use a mechanism similar to the Oldtimer API to read
+the metrics back.
+
+This would require every producer to know how to talk to every sink, would make
+configuring the sinks more complicated, and would most likely lead to software
+only being able to talk to one of the sinks supported by Heapster.
+Additionally, you lose the benefits of the Heapster model, and either have to
+adapt the existing Heapster model to fall back to an Oldtimer-like approach, or
+teach all cluster components to be able to use both the Heapster model and
+Oldtimer simultaneously.
+
+### Reworking the existing cAdvisor-Kubelet-Heapster Pull Mechanism
+
+This alternative would involve reworking the existing pull mechanism to allow
+certain pods to produce the "cluster-scoped" metrics (i.e. metrics scope to
+pods other than the producer), as opposed to the current situation, where all
+custom metrics collected via the current pull mechanism are scoped to the
+producer pod.
+
+This would require a mechanism for indicating to Heapster which pod names were
```

Comment:
With the main proposal, since producers already need to authenticate with Heapster, we get it for free -- you could potentially just say "whoever has the certificate that allows them to authenticate is allowed to push".  Since cAdvisor doesn't currently check certificates AFAIK, you could not use the same mechanism with cAdvisor -- you'd either have to teach cAdvisor to check certificates (which seems like tying cAdvisor closer to Kube than it wants to be), or set a list of pod names (or some sort of other identifier) at the Heapster level.